src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,880 @@
+                cost += (size / ptr_size) as usize;
+            } else {
+                cost += UNKNOWN_SIZE_COST;
+            }
+        }
+
+        debug!("Inline cost for {:?} is {}", callsite.callee, cost);
+
+        if let attr::InlineAttr::Always = hint {
+            true
+        } else {
+            cost <= threshold
+        }
+    }
+
+
+    fn inline_call(&self, callsite: CallSite<'tcx>,
+                             caller_mir: &mut Mir<'tcx>, callee_mir: Mir<'tcx>) -> bool {
+
+        // Don't inline a function into itself
+        if callsite.caller == callsite.callee { return false; }
+
+        let _task = self.tcx.dep_graph.in_task(DepNode::Mir(callsite.caller));
+
+
+        let terminator = caller_mir[callsite.bb].terminator.take().unwrap();
+        let cm = self.tcx.sess.codemap();
+        match terminator.kind {
+            // FIXME: Handle inlining of diverging calls
+            TerminatorKind::Call { args, destination: Some(destination), cleanup, .. } => {
+
+                debug!("Inlined {:?} into {:?}", callsite.callee, callsite.caller);
+
+                let is_box_free = Some(callsite.callee) == self.tcx.lang_items.box_free_fn();
+
+                let mut var_map = IndexVec::with_capacity(callee_mir.var_decls.len());
+                let mut temp_map = IndexVec::with_capacity(callee_mir.temp_decls.len());
+                let mut scope_map = IndexVec::with_capacity(callee_mir.visibility_scopes.len());
+                let mut promoted_map = IndexVec::with_capacity(callee_mir.promoted.len());
+
+                for mut scope in callee_mir.visibility_scopes {
+                    if scope.parent_scope.is_none() {
+                        scope.parent_scope = Some(callsite.location.scope);
+                        scope.span = callee_mir.span;
+                    }
+
+                    if !cm.is_valid_span(scope.span) {
+                        scope.span = callsite.location.span;
+                    }
+
+                    let idx = caller_mir.visibility_scopes.push(scope);
+                    scope_map.push(idx);
+                }
+
+                for mut var in callee_mir.var_decls {
+                    var.source_info.scope = scope_map[var.source_info.scope];
+
+                    if !cm.is_valid_span(var.source_info.span) {
+                        var.source_info.span = callsite.location.span;
+                    }
+                    let idx = caller_mir.var_decls.push(var);
+                    var_map.push(idx);
+                }
+
+                for temp in callee_mir.temp_decls {
+                    let idx = caller_mir.temp_decls.push(temp);
+                    temp_map.push(idx);
+                }
+
+                for p in callee_mir.promoted {
+                    let idx = caller_mir.promoted.push(p);
+                    promoted_map.push(idx);
+                }
+
+                // If the call is something like `a[*i] = f(i)`, where
```

Comment:
Mutating the lvalue of a call during it is UB anyway (read: trans does something random, and MIR construction never emits it). No need to handle it.